darkcluster/src/main/java/com/linkedin/darkcluster/impl/DarkClusterStrategyFactoryImpl.java

```diff
@@ -150,30 +150,34 @@ public void onClusterAdded(String updatedClusterName)
       // from the source cluster, and that called listenToCluster.
       // We also will be listening to updates on the dark clusters, because we'll be sending d2 requests
       // to the dark clusters, and will be listening on the dark cluster znodes.
-      // It is more precise to update on just dark cluster updates, because listening on the
-      // source cluster updates might have unrelated changes, whereas when a dark cluster update happens
-      // we know for sure we need to update that dark cluster.
-      if (_darkStrategyMap.containsKey(updatedClusterName))
+
+      // It should be sufficient to listen to updates on just the source cluster of the dark clusters, because all pertinent dark strategy info
+      // is contained in the source cluster entries. Source cluster changes will include stuff like dark multiplier changes and dark
+      // transportClientProperty changes, etc.
+      if (_sourceClusterName.equals(updatedClusterName))
       {
-        // this is a dark cluster name. however, to refresh the strategies, we need to pull the
-        // darkClusterConfigMap on the parent d2 cluster, because that has the properties needed
-        // to recreate the dark cluster strategies, such as the multiplier.
-        String darkClusterName = updatedClusterName;
         try
         {
-          DarkClusterConfigMap darkConfigMap = _facilities.getClusterInfoProvider().getDarkClusterConfigMap(_sourceClusterName);
-          if (darkConfigMap.containsKey(darkClusterName))
+          DarkClusterConfigMap potentialDarkConfigMap = _facilities.getClusterInfoProvider().getDarkClusterConfigMap(updatedClusterName);
+          for (Map.Entry<String, DarkClusterConfig> entry : potentialDarkConfigMap.entrySet())
           {
-            // just update the dark cluster that changed
-            _darkStrategyMap.put(darkClusterName, createStrategy(darkClusterName,
-                                                                 darkConfigMap.get(darkClusterName)));
+            // we need to refresh every dark cluster in this "source" cluster, because we don't have an easy way to tell
+            // what changed in the source cluster. It could have been a dark cluster change, or it could have been an unrelated
+            // change in the source cluster. Because there are so few changes in cluster znodes, this is not a
+            // big concern at the moment. If it does become a concern, storing the original darkClusterConfig and comparing with the
+            // new one is one way to reduce the strategy refreshes. However, since the strategy itself is so lightweight, it probably won't be a
+            // problem or worth the effort.
+            String darkClusterName = entry.getKey();
+            DarkClusterConfig darkClusterConfig = entry.getValue();
+            _darkStrategyMap.put(darkClusterName, createStrategy(darkClusterName, darkClusterConfig));
```

Comment:
made the changes we talked about. Also, I realized that the StrategyFactoryImpl's getOrCreate() shouldn't modify the strategyMap, because if there's constant traffic, it would make it impossible to remove a dark cluster strategy if the getOrCreate() added to the strategyMap.